lib/datadog/statsd/sender.rb

```diff
@@ -1,159 +1,130 @@
-        # tell sender-thread to notify us in the current
-        # thread's queue
-        message_queue.push(queue)
-        # wait for the sender thread to send a message
-        # once the flush is done
-        queue.pop
+      def flush(sync: false)
+        @mux.synchronize {
+          # we don't want to send a flush action to the bg thread if:
+          # - there is no bg thread running
+          # - the sender has been closed
+          return if !sender_thread.alive? || @is_closed
+
+          if sync
+             # blocking flush
+             blocking_queue = FlushQueue.new
+             channel << blocking_queue
+             blocking_queue.pop # wait for the bg thread to finish its work
+             blocking_queue.close if CLOSEABLE_QUEUES
+           else
+             # asynchronous flush
+             channel << :flush
+           end
+         }
       end
 
       def add(message)
-        raise ArgumentError, 'Start sender first' unless message_queue
+        return if @is_closed # don't send a message to the bg thread if the sender has been closed
 
-        # if the thread does not exist, we assume we are running in a forked process,
-        # empty the message queue and message buffers (these messages belong to
-        # the parent process) and spawn a new companion thread.
+        # the bg thread is not running anymore, this is happening if the main process has forked and
+        # we are running in the child, we will spawn a bg thread and reset buffers (containing parents' messages)
         if !sender_thread.alive?
-          @mx.synchronize {
-            # a call from another thread has already re-created
-            # the companion thread before this one acquired the lock
+          @mux.synchronize {
+            return if @is_closed
+            # test if a call from another thread has already re-created
+            # the background thread before this one acquired the lock
             break if sender_thread.alive?
-            @logger.debug { "Statsd: companion thread is dead, re-creating one" } if @logger
 
-            message_queue.close if CLOSEABLE_QUEUES
-            @message_queue = nil
-            message_buffer.reset
-            start
+            # re-create the channel of communication since we will spawn a new bg thread
+            channel.close if CLOSEABLE_QUEUES
+            @channel = Queue.new
+            message_buffer.reset # don't use messages appended by another fork
+            @sender_thread = Thread.new(&method(:send_loop))
           }
         end
 
-        message_queue << message
+        channel << message
       end
 
-      def start
-        raise ArgumentError, 'Sender already started' if message_queue
-
-        # initialize a new message queue for the background thread
-        @message_queue = Queue.new
-        # start background thread
-        @sender_thread = Thread.new(&method(:send_loop))
+      # Compatibility with `Sender`
+      def start()
       end
```

Comment:
With this change, neither `Sender` nor `SingleThreadedSender` use `#start`, so perhaps it would make sense to just remove them?